### ISSUE TITLE
Support libkrun (mostly for Virtio-GPU Venus), via krunkit

### ISSUE THREAD

**Issue by AkihiroSuda** [lima-vm/lima#2277]
libkrun seems to have merged the PR for supporting Virtio-GPU Venus (https://docs.mesa3d.org/drivers/venus.html)
- https://github.com/containers/libkrun/pull/174

QEMU still hasn't merged the patches for supporting Venus: https://patchew.org/Xen/20231219075320.165227-1-ray.huang@amd.com/

**Comment by AkihiroSuda** [comment]
lima should probably use https://github.com/containers/krunkit for implementing the libkrun driver.
Not sure if it should be a built-in driver, or we have to implement the plug-in system.

**Comment by balajiv113** [comment]
We can explore and finalise the approach for plug-in model. As i already see several experimental PR's for supporting different drivers. 

This will help us in providing different options based on performance and usecases.

**Comment by nirs** [comment]
See minikube krunkit driver:
https://github.com/kubernetes/minikube/pull/20826

**Comment by afbjorklund** [comment]
Would make it more sense to use the libkrun API instead of the "krunkit" CLI, like with the VZ API instead of the "vfkit" CLI?

https://github.com/containers/libkrun

**Comment by AkihiroSuda** [comment]
> Would make it more sense to use the libkrun API instead of the "krunkit" CLI, like with the VZ API instead of the "vfkit" CLI?
> 
> https://github.com/containers/libkrun

That should be also fine after finalizing the driver plugin gRPC, as libkrun uses Rust, not Go.

**Comment by afbjorklund** [comment]
The library is in C (i.e. the ABI is), so the Rust CLI wraps it in "unsafe"

https://github.com/containers/libkrun/blob/main/include/libkrun.h

Something like this code, which would work equally fine in Go too:

https://github.com/containers/krunkit/blob/main/src/context.rs

**Comment by afbjorklund** [comment]
One benefit of using libkrun, is that it also supports Linux (not only macOS)

This is good, so that it can be tested also on developer machines (unlike Vz)

---

Made a sample Go wrapper, found out that [cwrap](https://github.com/h12w/cwrap) doesn't support C99 `bool`...

```go
package main

import (
	"h12.io/cwrap"
)

var krun = &cwrap.Package{
	PacName: "krun",
	PacPath: "go-krun",
	From: cwrap.Header{
		Dir:           "/usr/local/include",
		File:          "libkrun.h",
		NamePattern:   `\Akrun(.*)`,
		CgoDirectives: []string{"pkg-config: libkrun"},
	},
}

func main() {
	err := krun.Wrap()
	if err != nil {
		panic(err)
	}
}
```

Then again, it was not impossible to fix. I guess **real C programmers** use `int`?

```go
package krun // import "github.com/containers/libkrun/go-krun"

func AddDisk(ctxId uint32, blockId string, diskPath string, readOnly bool) (ret int32)
func AddDisk2(ctxId uint32, blockId string, diskPath string, diskFormat uint32, ...) (ret int32)
func AddVirtiofs(ctxId uint32, cTag string, cPath string) (ret int32)
func AddVirtiofs2(ctxId uint32, cTag string, cPath string, shmSize int) (ret int32)
func AddVsockPort(ctxId uint32, port uint32, cFilepath string) (ret int32)
func AddVsockPort2(ctxId uint32, port uint32, cFilepath string, listen bool) (ret int32)
func CheckNestedVirt() (ret int32)
func CreateCtx() (ret int32)
func FreeCtx(ctxId uint32) (ret int32)
func GetShutdownEventfd(ctxId uint32) (ret int32)
func SetConsoleOutput(ctxId uint32, cFilepath string) (ret int32)
func SetDataDisk(ctxId uint32, diskPath string) (ret int32)
func SetEnv(ctxId uint32, envp []string) (ret int32)
func SetExec(ctxId uint32, execPath string, argv []string, envp []string) (ret int32)
func SetGpuOptions(ctxId uint32, virglFlags uint32) (ret int32)
func SetGpuOptions2(ctxId uint32, virglFlags uint32, shmSize int) (ret int32)
func SetGvproxyPath(ctxId uint32, cPath string) (ret int32)
func SetKernel(ctxId uint32, kernelPath string, kernelFormat uint32, initramfs string, ...) (ret int32)
func SetLogLevel(level uint32) (ret int32)
func SetMappedVolumes(ctxId uint32, mappedVolumes []string) (ret int32)
func SetNestedVirt(ctxId uint32, enabled bool) (ret int32)
func SetNetMac(ctxId uint32, cMac *byte) (ret int32)
func SetPasstFd(ctxId uint32, fd int32) (ret int32)
func SetPortMap(ctxId uint32, portMap []string) (ret int32)
func SetRlimits(ctxId uint32, rlimits []string) (ret int32)
func SetRoot(ctxId uint32, rootPath string) (ret int32)
func SetRootDisk(ctxId uint32, diskPath string) (ret int32)
func SetSmbiosOemStrings(ctxId uint32, oemStrings []string) (ret int32)
func SetSndDevice(ctxId uint32, enable bool) (ret int32)
func SetTeeConfigFile(ctxId uint32, filepath string) (ret int32)
func SetVmConfig(ctxId uint32, numVcpus byte, ramMib uint32) (ret int32)
func SetWorkdir(ctxId uint32, workdirPath string) (ret int32)
func Setgid(ctxId uint32, gid uint32) (ret int32)
func Setuid(ctxId uint32, uid uint32) (ret int32)
func SplitIrqchip(ctxId uint32, enable bool) (ret int32)
func StartEnter(ctxId uint32) (ret int32)
```

The documentation didn't make it either, so have to use libkrun.h for that.

----

EDIT: When comparing the auto-generated code, seems int was confused as well.

i.e. C `uint64_t` translated to Go `int`, but C `int` translated to Go `int32`. Oh well.

**Comment by nirs** [comment]
> One benefit of using libkrun, is that it also supports Linux (not only macOS)

libkrun on Linux is not the same as libkrun-efi for macOS. On Linux it is meant to run single application with better isolation. On macOS it can be used like vfkit to run standard vms.

So I think this will be a macOS only solution like vz and socket_vmnet.

**Comment by afbjorklund** [comment]
Yeah, I guess you would have to use external EFI - like you have to with QEMU

"libkrun-efi: Variant that bundles OVMF/EDK2 for booting a distribution-provided kernel"

Not sure about the single app though, the example runs a Fedora chroot shell

`./chroot_vm ./rootfs_fedora /bin/sh`

**Comment by afbjorklund** [comment]
> So I think this will be a macOS only solution like vz and socket_vmnet.

Makes sense, I think there are other drivers with better support for KVM

**QEMU with VirtIO GPU Vulkan Support** ([qemu virtio-gpu](https://www.qemu.org/docs/master/system/devices/virtio-gpu.html) virglrenderer)

https://gist.github.com/peppergrayxyz/fdc9042760273d137dddd3e97034385f

**Comment by afbjorklund** [comment]
Added the Go version here, with the basic libkrun wrapper and chroot_vm example:

https://github.com/afbjorklund/go-libkrun

https://github.com/containers/libkrun#chroot_vm

Fixed the [go-cwrap](https://github.com/afbjorklund/go-cwrap) program as per above, to handle stdbool and to not filter out int.

**Comment by afbjorklund** [comment]
Using libkrun from CLI instead of from API does cut down on the size of the dependencies, unless it's a completely separate driver

The main difference (to VZ) is that Hypervisor.framework and Virtualization.framework are system libraries, whereas libkrun is extra

**Comment by afbjorklund** [comment]
> QEMU still hasn't merged the patches for supporting Venus

Eventually the support got added, to QEMU version 9.2.0

**Comment by nirs** [comment]
> > QEMU still hasn't merged the patches for supporting Venus
> 
> Eventually the support got added, to QEMU version 9.2.0

On QEMU side it should support converting the Vulkan commands to Metal. Is this in qemu now?

**Comment by afbjorklund** [comment]
I think that is the job of [MoltenVK](https://github.com/KhronosGroup/MoltenVK), translating from Vulkan to Metal on the host side for macOS?

![Image](https://github.com/user-attachments/assets/cffe1a05-9b5e-4014-9216-d6d858834093)

* https://pretalx.com/kvm-forum-2024/talk/SVZZL9/ "The many faces of virtio-gpu"

Not sure what the status of it is in QEMU. Other platforms are using https://gitlab.freedesktop.org/virgl/virglrenderer

**Comment by nirs** [comment]
Yes libkrun uses moltenvk. I did not look how it does it, but I guess nobody implemented in qemu yet.

**Comment by afbjorklund** [comment]
It uses yet another project, called "rutabaga_gfx": https://docs.rs/rutabaga_gfx/latest/rutabaga_gfx/

Same as in https://www.qemu.org/docs/master/system/devices/virtio-gpu.html#virtio-gpu-rutabaga

**Comment by AkihiroSuda** [comment]
> qemu

See:
- https://github.com/lima-vm/lima/issues/2086

Seems to have got stuck due to lack of reviews :disappointed: :
- https://github.com/anholt/libepoxy/pull/239

**Comment by nirs** [comment]
Turns out that libkrun is compatible with socket_vment when using passt mode. The new API supporting multiple network interfaces will support socket_vment unix stream socket:
https://github.com/containers/libkrun/pull/367

I added feature request for krunkit to expose the new API:
https://github.com/containers/krunkit/issues/56

With this integrating krunkit should be easier. It can be a much simplified qemu driver that can work with user, shared, and bridged networks.

**Comment by nirs** [comment]
krunkit is adding support for socket_vmnet here:
https://github.com/containers/krunkit/pull/63

Tested here:
https://github.com/containers/krunkit/pull/63#issuecomment-3218148429

With this it should be pretty easy to add a driver that works like qemu, running krunkit instead of qemu. If add this it will be very easy to support also vfkit, since both have the same interface (krunkit was designed to be drop-in replacement for vfkit in podman).

**Comment by AkihiroSuda** [comment]
Anybody here planning to work on this?
@nirs @afbjorklund @balajiv113 ( and @unsuman )

Nice to have this as an experimental driver in Lima v2.0.

**Comment by afbjorklund** [comment]
I updated the go-libkrun wrapper, and included the go-cwrap patch (for stdbool) in case you decide to use it and to host it somewhere

**Comment by unsuman** [comment]
I'm actually working on it, sorry about not having a draft PR yet, but I'll get one up soon. It's taking a bit longer because I'm reverse engineering how Qemu and VZ work to figure out exactly what I need from the `store.Instance` struct to get VMs up and running with krunkit.

**Comment by paolomainardi** [comment]
> I'm actually working on it, sorry about not having a draft PR yet, but I'll get one up soon. It's taking a bit longer because I'm reverse engineering how Qemu and VZ work to figure out exactly what I need from the `store.Instance` struct to get VMs up and running with krunkit.

Great, and many thanks for your hard work.

Once you have something working, could you test this UID/GID scenario? It's related to this issue: https://github.com/lima-vm/lima/issues/4053, and it requires a Darwin system. If you do not have it, I would be more than happy to test it by myself on your branch.

**Test setup:**
```shell
❯ tree test-uid-gid
test-uid-gid
├── Dockerfile  
├── file
└── subdir
    └── subfile
```

**Expected behavior (working):**

```bash
❯ docker run -u 501 --rm -v $PWD:/test alpine:latest ash -c "ls -al / | grep test"
drwxr-xr-x    1 501      root           192 Sep 19 07:59 test

❯ docker run -u 501 --rm -v $PWD:/test alpine:latest ash -c "ls -al /test"  
drwxr-xr-x    1 501      root           192 Sep 19 07:59 .
drwxr-xr-x    1 root     root            20 Sep 19 10:17 ..
-rw-r--r--    1 501      root            47 Sep 19 07:59 Dockerfile
-rw-r--r--    1 501      root             0 Sep 19 06:58 file
drwxr-xr-x    1 501      root            96 Sep 19 08:00 subdir
```

**Current issue:**

Right now, the mount point doesn't preserve correct ownership when viewed from the parent directory:
- ✅ Inside `/test`: correct ownership (`501:root`)
- ❌ From root level: shows `root:root` instead of `501:root` 
- ✅ Files: correct ownership (`501:root`)

The key thing to verify is that the mount point itself maintains the correct UID when viewed from outside the mounted directory.

**Comment by afbjorklund** [comment]
Unfortunately libkrun works very differently on Linux, and not at all on Windows, so it is less portable than QEMU

**Comment by unsuman** [comment]
> Once you have something working, could you test this UID/GID scenario? It's related to this issue: https://github.com/lima-vm/lima/issues/4053, and it requires a Darwin system. If you do not have it, I would be more than happy to test it by myself on your branch.

Thanks! I have a Darwin System and will test it and let you know with Krunkit 😄

**Comment by paolomainardi** [comment]
Thanks a lot @unsuman

**Comment by paolomainardi** [comment]
Just an update on this issue https://github.com/lima-vm/lima/issues/4053 - I'm trying the same test case on a VM with libkrun through a podman machine, which defaults to this. The bug isn't present, so the issue appears to be somewhere in Apple's hypervisor/virtualization framework.

**Comment by AkihiroSuda** [comment]
@unsuman Do you have a WIP branch?

**Comment by unsuman** [comment]
> @unsuman Do you have a WIP branch?

I have successfully launched the VM using krunkit but was unable to setup networking and port forwarding. I'll raise the WIP PR and start work on networking in a day or two! Currently I dont have my laptop with me.

**Comment by AkihiroSuda** [comment]
Done in #4137 

Thanks @unsuman !

**Comment by paolomainardi** [comment]
Wow, this is huge! How can I test this ?

**Comment by AkihiroSuda** [comment]
> Wow, this is huge! How can I test this ?

Documented in https://lima-vm.io/docs/config/vmtype/krunkit/

**Comment by afbjorklund** [comment]
> Would make it more sense to use the libkrun API instead of the "krunkit" CLI, like with the VZ API instead of the "vfkit" CLI?

I archived the libkrun wrapper, now that we are going with the krunkit program. It was just using the C library anyway...
